### PR TITLE
Fixed #28905 - Added optional dependencies to extra_requires

### PR DESCRIPTION
**Pull request by jaap3**
https://code.djangoproject.com/ticket/28905

Make it easier to install Django including optional dependencies (like pyscopg2, Pillow etc.) with the correct minimum version by using, i.e.:

    pip install Django[postgresql,images]

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -51,6 +51,7 @@
     extras_require={
         "bcrypt": ["bcrypt"],
         "argon2": ["argon2-cffi >= 16.1.0"],
+        "postgresql": ["psycopg2>=2.5.4"]
```
Comment (sir-sigurd) [inline setup.py]: trailing comma?

**Comment by jaap3** [comment]
@pope1ni I like that even better, I was just thinking about my personal use case when I created this PR. I assume the change you propose probably has to go through a trac ticket.

**Comment by pope1ni** [comment]
As this is a little more significant a change, yes please.

File: setup.py [hunk 2]
```diff
@@ -49,8 +49,26 @@
...
     install_requires=['pytz'],
     extras_require={
-        "bcrypt": ["bcrypt"],
-        "argon2": ["argon2-cffi >= 16.1.0"],
+        'admindocs': ['docutils'],
+        'argon2': ['argon2_cffi>=16.1.0'],
+        'bcrypt': ['bcrypt'],
+        'gis': [
+            'geoip2',
+            'numpy'
```
Comment (sir-sigurd) [inline setup.py]: trailing comma?

File: setup.py [hunk 3]
```diff
@@ -49,8 +49,26 @@
...
+        ],
+        'geoip': ['geoip2'],
+        'jinja2': ['Jinja2>=2.9.2'],
+        'memcached': ['python3-memcached'],
+        'numpy': ['numpy'],
+        'mysql': ['mysqlclient>=1.3.7'],
+        'oracle': ['Oracle'],
```
Comment (sir-sigurd) [inline setup.py]: `cx_Oracle>=5.2`

https://docs.djangoproject.com/en/dev/ref/databases/#oracle-notes

File: setup.py [hunk 4]
```diff
@@ -49,8 +49,26 @@
...
+        'bcrypt': ['bcrypt'],
+        'gis': [
+            'geoip2',
+            'numpy'
+        ],
+        'geoip': ['geoip2'],
+        'jinja2': ['Jinja2>=2.9.2'],
+        'memcached': ['python3-memcached'],
+        'numpy': ['numpy'],
+        'mysql': ['mysqlclient>=1.3.7'],
```
Comment (sir-sigurd) [inline setup.py]: `mysqlclient>=1.3.3`

https://docs.djangoproject.com/en/dev/ref/databases/#id6

Comment (jaap3) [reply setup.py]: The test requirements specify >=1.3.7 (https://github.com/django/django/blob/master/tests/requirements/mysql.txt) "Due to a bug that will be fixed in mysqlclient 1.3.7."

I'm Not sure what's more correct.

File: setup.py [hunk 5]
```diff
@@ -49,8 +49,26 @@
...
+        'admindocs': ['docutils'],
+        'argon2': ['argon2_cffi>=16.1.0'],
+        'bcrypt': ['bcrypt'],
+        'gis': [
+            'geoip2',
+            'numpy',
+        ],
+        'geoip': ['geoip2'],
+        'jinja2': ['Jinja2>=2.9.2'],
+        'memcached': ['python3-memcached'],
```
Comment (pope1ni) [inline setup.py]: Just wanted to point out https://github.com/django/django/commit/770b9ea77fb5e39d616e62b54c06755e6d4f4d36 which was committed recently.

File: setup.py (hunk 4 above)
Comment (pope1ni) [reply setup.py]: See https://github.com/django/django/pull/9472 for a documentation change request.

File: setup.py (hunk 5 above)
Comment (carltongibson) [inline setup.py]: 1. Given that we have `pylibmc` below, is `memcached` the right key here?
2. Given that there are choices here, is this something we want to have an opinion on? (It doesn't seem as clear to me as e.g. databases.)

**Comment by timgraham** [comment]
Some documentation (which needs polish) was offered in https://github.com/django/django/pull/10124. Is it possible to avoid duplication between `setup.py` and `test/requirements/` (although I'm not they should necessarily always be the same?

**Comment by jaap3** [comment]
@timgraham I think it's possible to get rid of `test/requirements/` entirely. There could be an `'all'` extra target which would include all optional dependencies (which now is in `py3.txt`).

Installing dev Django for testing would then become something like `pip install -e .[postgresql,all]`

**Comment by pope1ni** [comment]
Superseded by #10663.

**Pull request closed without merging**